plugins/package-managers/bazel/src/main/kotlin/CompositeBazelModuleRegistryService.kt

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2024 The ORT Project Authors (see <https://github.com/oss-review-toolkit/ort/blob/main/NOTICE>)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.plugins.packagemanagers.bazel
+
+import java.io.File
+
+import org.apache.logging.log4j.kotlin.logger
+
+import org.ossreviewtoolkit.clients.bazelmoduleregistry.BazelModuleRegistryService
+import org.ossreviewtoolkit.clients.bazelmoduleregistry.LocalBazelModuleRegistryService
+import org.ossreviewtoolkit.clients.bazelmoduleregistry.ModuleMetadata
+import org.ossreviewtoolkit.clients.bazelmoduleregistry.ModuleSourceInfo
+import org.ossreviewtoolkit.clients.bazelmoduleregistry.RemoteBazelModuleRegistryService
+
+/**
+ * A composite Bazel module registry service that aggregates multiple [BazelModuleRegistryService] instances and
+ * delegates the package lookup to the corresponding registry. The first registry that contains the package is used.
+ * [packagesPerRegistry] is a map of [BazelModuleRegistryService] instances to the list of packages they are
+ * responsible for.
+ */
+class CompositeBazelModuleRegistryService(
+    private val packagesPerRegistry: Map<BazelModuleRegistryService, List<String>>
+) : BazelModuleRegistryService {
+    companion object {
+        // A regular expression to extract the server and package parts of a registry URL.
+        internal val URL_REGEX = "^(?<server>.*/)modules/(?<package>[^/]+)/[^/]+/source\\.json$".toRegex()
+
+        /**
+         * Create a Composite Bazel Module Registry client instance.
+         */
+        fun create(urls: Set<String>, projectDir: File): CompositeBazelModuleRegistryService {
```

Comment:
I think `URL_REGEX` should not be unit tested directly. Instead factoring out the following would allow
for testing a bit more logic and less details:
```
fun registryForUrl(urls: Set<String>): Map<BazelModuleRegistryService, Set<String>>`
```
Not sure if it would help to test remote and local separately, e.g.:
```
fun remoteRegistriesFor(urls: Set<String>): Map<BazelModuleRegistryService, Set<String>>`
fun localRegistriesFor(urls: Set<String>): Map<BazelModuleRegistryService, Set<String>>`
```

Reply:
I disagree: testing the regex directly allow you to find _quickly_ errors in it, which is the critical part of the service: this is the location where it takes outside world data and convert it to the internal representation.
This quickness is IMO the goal of a unit test.

If you _replace_ it by testing `registryForUrl`, your test output in case of failure would be: "some registries are missing". Then, you would have to take some time finding out why.

However, I agree that `registryForUrl` should also be tested. Can we keep it for a future commit?

Reply:
> If you replace it by testing registryForUrl, your test output in case of failure would be: "some registries are missing". Then, you would have to take some time finding out why.

It wouldn't need to be tested for 100eds of registries, but maybe with 1,2,3 URLs.
So, it would not be difficult IMO to understand the test.

Anyhow, let's jst disagree and continue.